lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php

```diff
@@ -204,48 +204,67 @@ public function connect($connectionName = null)
      */
     protected function connectTo($connectionName)
     {
+        return $connectionName == 'master' ?
+            $this->connectToMaster() :
+            $this->connectToSlave();
+    }
+
+    /**
+     * Connets to a master
+     *
+     * @return Driver\Connection
+     *
+     * @throws ConnectionException
+     */
+    private function connectToMaster()
+    {
         $params = $this->getParams();
 
         $driverOptions = isset($params['driverOptions']) ? $params['driverOptions'] : array();
 
-        $connectionParams = $this->chooseConnectionConfiguration($connectionName, $params);
+        $connectionParams = $params['master'];
 
         $user = isset($connectionParams['user']) ? $connectionParams['user'] : null;
         $password = isset($connectionParams['password']) ? $connectionParams['password'] : null;
 
-        if ($connectionName == 'slave') {
-            while(true) {
-                try {
-                    return $this->_driver->connect($connectionParams, $user, $password, $driverOptions);
-                } catch (ConnectionException $e) {
-                    if (count($params['slaves']) <= 0) {
-                        throw new ConnectionException('No slaves are available', $e->getPrevious());
-                    }
-
-                    $connectionParams = array_pop($params['slaves']);
-
-                    $user = isset($connectionParams['user']) ? $connectionParams['user'] : null;
-                    $password = isset($connectionParams['password']) ? $connectionParams['password'] : null;
-                }
-            }
-        } else {
-            return $this->_driver->connect($connectionParams, $user, $password, $driverOptions);
-        }
+        return $this->_driver->connect($connectionParams, $user, $password, $driverOptions);
     }
 
     /**
-     * @param string $connectionName
-     * @param array  $params
+     * Connets to a random slave and tries others until if found a server that is alive
      *
-     * @return mixed
+     * @return Driver\Connection
+     *
+     * @throws ConnectionException
      */
-    protected function chooseConnectionConfiguration($connectionName, $params)
+    private function connectToSlave()
     {
-        if ($connectionName === 'master') {
-            return $params['master'];
+        $params = $this->getParams();
+
+        $driverOptions = isset($params['driverOptions']) ? $params['driverOptions'] : array();
+
+        shuffle($params['slaves']);
+
+        do {
+            $connectionParams = array_pop($params['slaves']);
```

Comment:
If you design it with this approach, then the loop can be `while ($connectionParams = array_pop($params['slaves'])) {`